src/bases/createWidgetBase.ts

```diff
@@ -0,0 +1,201 @@
+import { ComposeFactory } from 'dojo-compose/compose';
+import createStateful from 'dojo-compose/bases/createStateful';
+import {
+	HNode,
+	DNode,
+	WNode,
+	Widget,
+	WidgetMixin,
+	WidgetState,
+	WidgetOptions
+} from 'dojo-interfaces/widgetBases';
+import { VNode, VNodeProperties } from 'dojo-interfaces/vdom';
+import { Factory } from 'dojo-interfaces/core';
+import { assign } from 'dojo-core/lang';
+import WeakMap from 'dojo-shim/WeakMap';
+import Map from 'dojo-shim/Map';
+import d from './../util/d';
+
+export interface WidgetFactory extends ComposeFactory<Widget<WidgetState>, WidgetOptions<WidgetState>> {}
+
+interface WidgetInternalState {
+	readonly id?: string;
+	dirty: boolean;
+	widgetClasses: string[];
+	cachedWidget?: VNode;
+	historicChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+	currentChildrenMap: Map<string | Factory<Widget<WidgetState>, WidgetOptions<WidgetState>>, Widget<WidgetState>>;
+};
+
+/**
+ * Internal state map for widget instances
+ */
+const widgetInternalStateMap = new WeakMap<Widget<WidgetState>, WidgetInternalState>();
+
+/**
+ * The counter for generating a unique ID
+ */
+let widgetCount = 0;
+
+function generateID(instance: Widget<WidgetState>): string {
+	const id = `widget-${++widgetCount}`;
+	instance.setState({ id });
+	return id;
+}
+
+function isWNode(child: DNode): child is WNode {
+	return child && (<WNode> child).factory !== undefined;
+}
+
+function realizeDNode(instance: Widget<WidgetState>, dNode: DNode) {
+	const internalState = widgetInternalStateMap.get(instance);
+	let child: HNode | Widget<WidgetState>;
+	if (isWNode(dNode)) {
+		const { factory, options: { id, state } } = dNode;
+		const childrenMapKey = id || factory;
+		const cachedChild = internalState.historicChildrenMap.get(childrenMapKey);
+		if (cachedChild) {
+			child = cachedChild;
+			if (state) {
+				child.setState(state);
+			}
+		} else {
+			child = factory(dNode.options);
+			child.own(child.on('invalidate', () => {
```

Comment:
you'd want the `child` to own it also no? If the child was destroyed then you wouldn't want the listener hanging around? If the `instance` is destroyed then the child will implicitly be destroyed and therefore the listener also.